modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java

```diff
@@ -0,0 +1,543 @@
+            }
+            // Create a listed resource object out of the returned users list.
+            ListedResource listedResource = createListedResource(returnedRoles, searchRequestObject.getStartIndex(),
+                    totalResults);
+            String encodedListedResource = encoder.encodeSCIMObject(listedResource);
+            Map<String, String> responseHeaders = new HashMap<>();
+            responseHeaders.put(SCIMConstants.CONTENT_TYPE_HEADER, SCIMConstants.APPLICATION_JSON);
+            return new SCIMResponse(ResponseCodeConstants.CODE_OK, encodedListedResource, responseHeaders);
+
+        } catch (CharonException | InternalErrorException | BadRequestException | NotImplementedException e) {
+            return AbstractResourceManager.encodeSCIMException(e);
+        }
+    }
+
+    @Override
+    public SCIMResponse updateWithPUTRole(String id, String putRequest, RoleManager roleManager) {
+
+        try {
+            if (roleManager == null) {
+                String error = "Provided role manager is null.";
+                throw new InternalErrorException(error);
+            }
+            JSONEncoder encoder = getEncoder();
+            JSONDecoder decoder = getDecoder();
+            SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getRoleResourceSchema();
+            Map<String, Boolean> requestAttributes = ResourceManagerUtil.getAllAttributeURIs(schema);
+            Role role = decoder.decodeResource(putRequest, schema, new Role());
+            Role updatedRole;
+
+            // Retrieve the old object.
+            Role oldRole = roleManager.getRole(id, requestAttributes);
+            if (oldRole != null) {
+                Role newRole = (Role) ServerSideValidator.validateUpdatedSCIMObject(oldRole, role, schema);
+                updatedRole = roleManager.updateRole(oldRole, newRole);
+            } else {
+                String error = "No role exists with the given id: " + id;
+                throw new NotFoundException(error);
+            }
+            return getScimResponse(encoder, updatedRole);
+
+        } catch (NotFoundException | BadRequestException | CharonException | ConflictException | InternalErrorException
+                | NotImplementedException e) {
+            return encodeSCIMException(e);
+        }
+    }
+
+    @Override
+    public SCIMResponse updateWithPATCHRole(String id, String patchRequest, RoleManager roleManager) {
+
+        try {
+            if (roleManager == null) {
+                String error = "Provided role manager handler is null.";
+                throw new InternalErrorException(error);
+            }
+            JSONEncoder encoder = getEncoder();
+            SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getRoleResourceSchema();
+            Map<String, Boolean> requestAttributes = ResourceManagerUtil.getAllAttributeURIs(schema);
+
+            Role oldRole = roleManager.getRole(id, requestAttributes);
```

Comment:
will this retrieve the entire user list and group list of role?

Reply:
Yes. since it needed to compare.

Reply:
let's create a git issue to track the perf improvement